### PR TITLE
Fix stack overflow reporting with multiple PALs

### LINKED ISSUE #107226: Secondary frame stackoverflow test failure on linux
This test:
https://github.com/dotnet/runtime/blob/ac663e44b907618c631edd614231b73677843952/src/tests/baseservices/exceptions/stackoverflow/stackoverflowtester.cs#L157

is failing in `runtime-coreclr superpmi-collect-test` pipeline on linux x64, arm64 and arm:

```console
Running stackoverflow test(largeframe main)
"Stack overflow."
"Repeated 42 times:"
"--------------------------------"
"   at TestStackOverflow.Program.InfiniteRecursionB2()"
"   at TestStackOverflow.Program.InfiniteRecursionA2()"
"   at TestStackOverflow.Program.InfiniteRecursionC2()"
"--------------------------------"
"   at TestStackOverflow.Program.InfiniteRecursionB2()"
"   at TestStackOverflow.Program.InfiniteRecursionA2()"
"   at TestStackOverflow.Program.Test(Boolean)"
"   at TestStackOverflow.Program.Main(System.String[])"
""
Running stackoverflow test(smallframe secondary)
"Stack overflow."
"Stack overflow."
""
System.Exception: Missing "TestStackOverflow.Program.Test" method frame
   at TestStackOverflow.Program.TestStackOverflowSmallFrameSecondaryThread()
   at __GeneratedMainWrapper.Main()
Expected: 100
Actual: 101
END EXECUTION - FAILED
```

cc @janvorli

### PR DESCRIPTION
**Pull request by janvorli**
A recently reported issue in the stack overflow test running with superpmi collect has revealed that there is a problem when multiple shared libraries using coreclr PAL are loaded into the process. Each such library registers signal handlers, including the SIGSEGV one. The unexpected thing in the SIGSEGV handler in this scenario is the fact that the GetCurrentPalThread() returns NULL in the non-coreclr shared libraries in case the library didn't invoke any function that would create the thread object.
That leads to the handler just writing "Stack overflow." to the console and then calling PROCAbort(). This abort causes the process to be torn down without reporting the stack overflow with full stack trace by the coreclr.

This change fixes it by calling the previous registered handler instead aborting in this case. That gives the coreclr SIGSEGV handler a chance to do the reporting as expected.

Close #107226

### CODE REVIEW

**Comment by am11** [comment]
Thank you! Please execute `/azp run runtime-coreclr superpmi-collect-test`.

**Comment by jakobbotsch** [comment]
#107277 should fix superpmi-collect-test.

**Comment by jakobbotsch** [comment]
All runtime tests passed though, so seems like this change fixes the problem.

**Comment by am11** [comment]
Yes, it was failing on x64 as well, which is now green. 👍

File: src/coreclr/pal/src/exception/signal.cpp [hunk 1]
```diff
@@ -653,7 +653,6 @@ static void sigsegv_handler(int code, siginfo_t *siginfo, void *context)
             else
```
Comment (jkotas) [inline src/coreclr/pal/src/exception/signal.cpp:653]: Nit: The "The current executable (shared library) doesn't have...." comment can be moved after the `if ... else `. It applies to the else part too now.

**Comment by jkotas** [comment]
cc @lambdageek Aleksey has been looking into removing superpmi dependency on PAL exception handling.

**Review by jkotas — APPROVED** [review]

File: src/coreclr/pal/src/exception/signal.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/pal/src/exception/signal.cpp:650]: Makes sense, moved.

**Comment by lambdageek** [comment]
> cc @lambdageek Aleksey has been looking into removing superpmi dependency on PAL exception handling.

:+1: My idea is to remove the `CatchHardwareExceptionHandler` mechanism from the PAL and add a sigsegv singnal handler to superpmi.  (And as a technical simplification instead of turning a SIGSEGV into a C++ exception that is then caught by superpmi to run their "finally" logic, it will instead wake a utility thread that will run the superpmi "finally" code).  

Like the changes here, by SIGSEGV handler will chain to the previous one, if it is present, rather than abort

**Pull request merged**